adjust_binning.py

```diff
@@ -0,0 +1,565 @@
+                            path_found = False
+                            break
+                        temp_config = temp_config[split_var]
+                        if not (isinstance(temp_config, dict) and key in temp_config):
+                            break
+                        temp_config = temp_config[key]
+
+                    if not (path_found and isinstance(temp_config, (int, float, list, tuple))):
+                        continue  # no valid value
+
+                    value_to_add = temp_config
+
+                    if not isinstance(value_to_add, (list, tuple)):
+                        value_to_add = [value_to_add]
+
+                    bins = value_to_add + bins if option == "add_left" else bins + value_to_add
+
+                events_per_bin = []
+                if len(bins) > 1:
+                    pairs = list(pairwise(bins))
+                    for lower_edge, upper_edge in pairs[:-1]:
+                        count = sub_df.query(f"{lower_edge} <= {target_variable} < {upper_edge}").shape[0]
+                        events_per_bin.append(count)
+
+                    lower_edge, upper_edge = pairs[-1]  # exclusive lower edge, inclusive upper edge
+                    count = sub_df.query(f"{lower_edge} <= {target_variable} <= {upper_edge}").shape[0]
+                    events_per_bin.append(count)
+
+                table.add_row(f"{' ' * prepend_space}{cat_key}", str(len(sub_df)), str(bins), str(events_per_bin))
+                output_bins[cat_key] = bins
+            else:
+                table.add_row(f"{' ' * prepend_space}{cat_key}", str(len(sub_df)), "No binning", "")
+
+    console.print(table)
+
+    return output_bins, parent_cat_keys, generated_edges, generated_categories
+
+
+if __name__ == "__main__":
+    args = parser.parse_args()
+    args.config = Path(args.config).resolve()
+
+    console = Console()
+    console.print(f"Loading configuration from [cyan]{args.config}[/cyan]")
+    console.print(f"Processes to adjust binning for: [cyan]{args.processes}[/cyan]")
+    console.print(f"Cut region: [cyan]{args.cut_region}[/cyan]")
+
+    cuts_config = None
+    if args.cuts_config:
+        console.print(f"Loading cuts from [cyan]{args.cuts_config}[/cyan]")
+        with open(Path(args.cuts_config).resolve(), "r") as f:
+            cuts_config = func.configured_yaml.load(f)
+
+    with open((args.config.parent / "common_settings.yaml").resolve(), "r") as f:
+        common_settings = func.configured_yaml.load(f)
+
+    with open(args.config, "r") as f:
+        config = func.configured_yaml.load(f)
+
+    base_directory, directory, era, channel, tree, file = (
+        Path(common_settings["output_path"]),
+        "preselection",
+        common_settings["era"],
+        config["channel"],
+        common_settings["tree"],
+        "data.root",
+    )
+    root_file_path = base_directory / directory / era / channel / file
```

Comment:
This could be shortened, no? None of these variables are used later, besides channel once.